### PR TITLE
server: fix sub query error in prepare statement

### PR DESCRIPTION
**Pull request by jackysp**
When prepare a statement with sub queries by api, TiDB will throw "transaction future is not set". This pr fix it. PTAL @tiancaiamao @coocood

### CODE REVIEW

**Comment by coocood** [comment]
LGTM

File: session.go [hunk 1]
```diff
@@ -836,8 +838,12 @@ func (s *session) PrepareStmt(sql string) (stmtID uint32, paramCount int, fields
 		return
 	}
 
+	s.PrepareTxnCtx(goCtx)
 	prepareExec := executor.NewPrepareExec(s, executor.GetInfoSchema(s), sql)
 	err = prepareExec.DoPrepare()
+	if err == nil {
+		err = s.doCommit(goCtx)
```
Comment (tiancaiamao) [inline session.go]: Why `doCommit` here?
What's behavior of MySQL?  

```
begin;
insert into t values (1);
prepare stmt from 'select * from t';
rollback;
```

Will 1 be inserted?

File: session.go [hunk 2]
```diff
@@ -836,6 +836,7 @@ func (s *session) PrepareStmt(sql string) (stmtID uint32, paramCount int, fields
 		return
 	}
 
+	s.PrepareTxnCtx(goctx.Background())
```
Comment (tiancaiamao) [inline session.go]: What's the timestamp used when the prepared statement execute?
Why compile a prepare statement need timestamp?
@jackysp @hanfei1991

File: session.go (hunk 1 above)
Comment (jackysp) [reply session.go]: fixed.

**Comment by coocood** [comment]
@jackysp 
If after prepare, the table changed, then we execute the prepared statement, we should see the new value, please write a test for this case.

File: session.go (hunk 2 above)
Comment (jackysp) [reply session.go]: Please see this pr https://github.com/pingcap/tidb/pull/5116 for some details. @tiancaiamao

**Comment by shenli** [comment]
@jackysp Please address the comment.

File: session.go [hunk 3]
```diff
@@ -836,10 +836,24 @@ func (s *session) PrepareStmt(sql string) (stmtID uint32, paramCount int, fields
 		return
 	}
 
-	s.PrepareTxnCtx(goctx.Background())
+	goCtx := goctx.Background()
+	inTxn := s.GetSessionVars().InTxn()
+	isAutocommit := s.GetSessionVars().IsAutocommit()
```
Comment (coocood) [inline session.go]: I think we don't need to consider auto-commit.
We only care about  `inTxn`.
prepare should not start a new transaction.

Comment (jackysp) [reply session.go]: You are right. Prepare statement should not start a new transaction. Fixed PTAL

**Comment by jackysp** [comment]
PTAL @wentaoxu

**Comment by coocood** [comment]
LGTM

File: executor/prepared_test.go [hunk 4]
```diff
@@ -70,8 +71,30 @@ func (s *testSuite) TestPrepared(c *C) {
...
+		rows, err := tidb.GetRows4Test(goCtx, tk.Se, rs)
+		c.Assert(err, IsNil)
+		c.Assert(rows, NotNil)
+		c.Assert(rs.Close(), IsNil)
+
+		tk.MustExec("insert prepare_test (c1) values (2),(2),(NULL)")
+		rs, err = tk.Se.ExecutePreparedStmt(goCtx, stmtId, 2)
+		rows, err = tidb.GetRows4Test(goCtx, tk.Se, rs)
+		c.Assert(err, IsNil)
+		c.Assert(rows, NotNil)
```
Comment (tiancaiamao) [inline executor/prepared_test.go]: I think it's better to check values rather than NotNil

File: executor/prepared_test.go [hunk 5]
```diff
@@ -70,8 +71,30 @@ func (s *testSuite) TestPrepared(c *C) {
...
+		tk.MustExec("begin")
+		tk.MustExec("insert prepare_test (c1) values (3),(2),(NULL)")
+		query = "select c1, c2 from prepare_test where c1 = ?"
+		stmtId, _, _, err = tk.Se.PrepareStmt(query)
 		c.Assert(err, IsNil)
+		tk.MustExec("rollback")
+		rs, err = tk.Se.ExecutePreparedStmt(goCtx, stmtId, 3)
+		rows, err = tidb.GetRows4Test(goCtx, tk.Se, rs)
+		c.Assert(err, IsNil)
+		c.Assert(rows, IsNil)
```
Comment (tiancaiamao) [inline executor/prepared_test.go]: ditto

File: session.go [hunk 6]
```diff
@@ -836,9 +836,23 @@ func (s *session) PrepareStmt(sql string) (stmtID uint32, paramCount int, fields
 		return
 	}
 
+	goCtx := goctx.Background()
+	inTxn := s.GetSessionVars().InTxn()
+	s.PrepareTxnCtx(goCtx)
```
Comment (tiancaiamao) [inline session.go]: Add comment:
// NewPrepareExec may need startTS to build the executor, so we have to call PrepareTxnCtx here.

File: session.go [hunk 7]
```diff
@@ -836,9 +836,23 @@ func (s *session) PrepareStmt(sql string) (stmtID uint32, paramCount int, fields
...
 	prepareExec := executor.NewPrepareExec(s, executor.GetInfoSchema(s), sql)
 	err = prepareExec.DoPrepare()
-	return prepareExec.ID, prepareExec.ParamCount, prepareExec.Fields, errors.Trace(err)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	if !inTxn {
+		err = s.CommitTxn(goCtx)
```
Comment (tiancaiamao) [inline session.go]: Rollback is better

File: session.go [hunk 8]
```diff
@@ -836,9 +836,23 @@ func (s *session) PrepareStmt(sql string) (stmtID uint32, paramCount int, fields
...
+	inTxn := s.GetSessionVars().InTxn()
+	s.PrepareTxnCtx(goCtx)
 	prepareExec := executor.NewPrepareExec(s, executor.GetInfoSchema(s), sql)
 	err = prepareExec.DoPrepare()
-	return prepareExec.ID, prepareExec.ParamCount, prepareExec.Fields, errors.Trace(err)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	if !inTxn {
```
Comment (tiancaiamao) [inline session.go]: Add comment:
// As we start a fake transaction to build the prepare executor before, we should rollback it here.

File: session.go (hunk 7 above)
Comment (tiancaiamao) [reply session.go]: rest looks good to me

**Comment by jackysp** [comment]
PTAL @tiancaiamao

File: executor/prepared_test.go [hunk 9]
```diff
@@ -72,29 +71,27 @@ func (s *testSuite) TestPrepared(c *C) {
 		stmtId, _, _, err := tk.Se.PrepareStmt(query)
 		c.Assert(err, IsNil)
 		rs, err := tk.Se.ExecutePreparedStmt(goCtx, stmtId, 1)
-		rows, err := tidb.GetRows4Test(goCtx, tk.Se, rs)
 		c.Assert(err, IsNil)
-		c.Assert(rows, NotNil)
```
Comment (tiancaiamao) [inline executor/prepared_test.go]: We can't check the row values directly?

File: util/testkit/testkit.go [hunk 10]
```diff
@@ -170,6 +170,12 @@ func (tk *TestKit) MustQuery(sql string, args ...interface{}) *Result {
 	rs, err := tk.Exec(sql, args...)
 	tk.c.Assert(errors.ErrorStack(err), check.Equals, "", comment)
 	tk.c.Assert(rs, check.NotNil, comment)
+	return tk.ResultsToStr(rs, comment)
+}
+
+// ResultsToStr converts ast.RecordSet to testkit.Result.
+// It is used to check results of execute statement in binary mode.
+func (tk *TestKit) ResultsToStr(rs ast.RecordSet, comment check.CommentInterface) *Result {
```
Comment (tiancaiamao) [inline util/testkit/testkit.go]: How about name it `RecordSetToResult`, the return value is not a string...

File: executor/prepared_test.go [hunk 11]
```diff
@@ -70,8 +70,28 @@ func (s *testSuite) TestPrepared(c *C) {
 		query := "select c1, c2 from prepare_test where c1 = ?"
 		stmtId, _, _, err := tk.Se.PrepareStmt(query)
 		c.Assert(err, IsNil)
-		_, err = tk.Se.ExecutePreparedStmt(goCtx, stmtId, 1)
+		rs, err := tk.Se.ExecutePreparedStmt(goCtx, stmtId, 1)
```
Comment (tiancaiamao) [inline executor/prepared_test.go]: Does it help?
`tk.MustQuery("select c1, c2 from prepare_test where c1 = ?", 1)`

It's basically the same with your prepared statement.

Comment (jackysp) [reply executor/prepared_test.go]: I just add a result check for this. I think these code are used to call prepare and execute directly.

File: util/testkit/testkit.go (hunk 10 above)
Comment (jackysp) [reply util/testkit/testkit.go]: It is basically a string array with comment. Anyway I'll change the name.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**